### DISCUSSION TITLE
Should `StaticInt` be in Base?

### DISCUSSION THREAD

**Discussion by Tokazama** [JuliaLang/julia#44519] [General]
Would `struct StaticInt{N::Int} <: Integer end` be a welcome addition here? Most of what it would involve is found [here](https://github.com/SciML/Static.jl/blob/master/src/int.jl).

It's helpful to have it specifically be a subtype of `Integer` so that it can provide optionally static ranges without the user having to worry about explicitly propagating static information from something like `SArray` or making the `dim` argument of something like `f(A, dim)` type stable. There are also times it can reduce the number of unique types generated for trait systems. For example [`contiguous_axis`](https://github.com/JuliaArrays/ArrayInterface.jl/blob/b2889edbdfc09230d3a390881f959013eac3e1b2/src/stridelayout.jl#L103) returns a `StaticInt` instead of `ContiguousAxis{N}`. The benefit of having it in base would be better native support, integration, and propagation of types/traits that use `StaticInt`

I understand there may be some concern that this overextends the intended use of the type system, and instead the focus should be better constant propagation on the compiler side of things so that something like `SArray` doesn't need this. I'm not sure if the other use cases are also better handled by some future compiler wizardry but I thought I'd at least put it forward.

EDIT:
If we ultimately don't go forward with this because the utility of `StaticInt` will be replaced with more appropriate compiler optimizations, it would be nice to have some clarity on how that will effect code development moving forward. For example, `Base.OneTo` indexing another array is explicitly written to only check the upper bound is in bounds. `StaticArrays.SOneTo` does everything in the type domain so nothing is done at run time. `ArrayInterface.OptionallyStaticUnitRange` uses `Int` or `StaticInt` to potentially allow either of those optimizations. Will optimizations to the compiler allow the same benefits without using type information. In other words, will the compiler turn [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1143) into [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1140)? If that's not an anticipated benefit of compiler work then how do we appropriately approach problems like these without abusing the type system?

@chriselrod is familiar with the `StaticInt` implementation in `Static.jl` and may have more to add to the discussion.
I think @timholy at one point expressed some support for this being in base.
@Keno probably has an opinion on whether this will eventually not be useful to maintain in base given more internal development.

**Comment by chriselrod** [comment]
> I understand there may be some concern that this overextends the intended use of the type system, and instead the focus should be better constant propagation on the compiler side of things so that something like SArray doesn't need this. I'm not sure if the other use cases are also better handled by some future compiler wizardry but I thought I'd at least put it forward.

Heavy use of Static.jl was cited as one of the reasons for LoopVectorization.jl's bad compile times.
Currently, LoopVectorization.jl does a lot of its code generation through leveraging Julia's type system (both in dispatches, and in using `StaticInt` to tell `@generated` functions inside VectorizationBase what they should be doing), when it'd be better off doing all of its code generation itself.

Given that latency is perhaps the number one problem with Julia, I do think people should be wary of investing too much in this approach, as they're likely to suffer consequences when they take it far.
I am working on getting a replacement to LoopVectorization that avoids these issues and does not use Static.jl.

**Comment by cscherrer** [comment]
@chriselrod it's important to avoid problems like this, but this approach sounds a little like reinventing the wheel. Since it comes up so often, wouldn't it be better to have a reusable solution, so we don't have to solve each instance independently? I might well be misunderstanding, since I haven't seen details of the replacement you're working on.

**Comment by Tokazama** [comment]
If the replacement completely removes the need to encode the size at the type level for something like `SArray` that would certainly be helpful but even then I'm assuming these compiler passes would only apply to specific array operations. We'd still have a bunch of use cases for things like `Base.OneTo` where we want to optimize on some known information and it is encoded at the type level. Or perhaps even that is solvable with future development?

**Comment by chriselrod** [comment]
Sorry, to be clear, it's a replacement for LoopVectorization.
What I was trying to warn is, someone going down that road may find themselves rewriting later to avoid these problems.

**Comment by AriMKatz** [comment]
The replacement is to rewrite in c++ IIUC, right?  Presumably we'd like to have  whatever it is that makes C++ more amenable to this, In Julia

**Reply by chriselrod** [reply]
That part (generating `Expr` or some IR without using the type system to do it) could of course be done in Julia as well.

There's lots of other reasons to favor C++ for this, some more easily addressable (e.g. as poorly as LLVM is documented, it is still much better documented than Julia's compiler) than others (e.g. wanting to run as late as possible in the compiler pipeline, when code is at its cleanest/simplest point).

**Comment by Tokazama** [comment]
If it's decided this shouldn't be in base I'm hoping to get an idea of where the rest of the community should be going with things like traits and corresponding optimizations. Compiler work is super important but I'm worried about requiring any similar additions in the future to be through the compiler. Perhaps this will change, but that sort of coding seems like it requires a lot of extra knowledge to do right.

**Reply by AriMKatz** [reply]
The other thing about compiler work is that sometimes it's unclear how to hit the right semantics to get the compiler to optimize stuff, tbecause we don't have static typing or when what we ask of the compiler has outrun the semantics of the language( which were initially co-designed with a much less demanding set of expectations, but now there are GPUs ad, etc )

**Reply by AriMKatz** [reply]
Also, this isn't strictly necessary, but the nice thing about having things in the type domain is that it helps with correctness

**Comment by AriMKatz** [comment]
This is also important for things like https://github.com/JuliaLang/julia/issues/43642 and related compiler optimizations. 

@Tokazama What about named dimensions or ragged arrays?

xref discussion here https://discourse.julialang.org/t/is-my-understanding-of-julia-correct/76924/33

**Comment by Tokazama** [comment]
I'm sure there are other collections than dense arrays that could be optimized for some operations and it would be a shame if that could only ever happen in the compiler. If we don't need to worry about static values then it would be great to have named dimensions just be a `Symbol` and not worry about them at the type level, but I'm still not sure how that would work. How do we go past constant propagation and follow something throughout the entire call graph for optimization?

I'm not saying that using `StaticInt` solves these problems, but it seems to be a good user facing way of ensuring a value is statically known to the compiler. A lot of use cases of `Val` can be replaced with a `StaticInt`, and a proper subtype is helpful in those values propagating through other functions without writing extra code.

**Comment by AriMKatz** [comment]
Agreed. I highly recommend you check out https://youtu.be/npDCCVIaSVQ  to see how a from scratch re-design of array typing solves these issues. I wonder how much we can adapt from that into julia. 

There are also the issues of making these arrays play nicely with GPU and AD (and as a bonus, shape checking) , which Dex's type system takes into account

**Reply by Tokazama** [reply]
That video is pretty interesting and I can see how Julia has similar solutions for some parts (e.g., using `eachindex(A)` to produce `OneTo` where we don't need to check for one based indexing). I'm going to update my first comment with some of that in perspective.

**Comment by timholy** [comment]
It's a tough call. But I guess I fall into the camp of thinking that since it's not yet been made unnecessary by compiler optimizations, it seems to be asking for trouble to trust that it will (soon) be. For that reason, I'd suggest we submit this to Base, and give people until a month or so before the 1.9 feature freeze to implement compiler optimizations that make it unnecessary. If that doesn't happen, then I say we merge it (along with other traits that have been found generically useful).

**Reply by Tokazama** [reply]
[Here's](https://github.com/JuliaLang/julia/pull/44538) the PR.

**Comment by JeffBezanson** [comment]
> will the compiler turn [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1143) into [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1140)?

I don't think that specifically could happen, since the compiler cannot switch a `UnitRange` to a `OneTo`. But, it can constant propagate it:

```
julia> f() = intersect(1:3, 2:4)

julia> @code_typed f()
CodeInfo(
1 ─     return $(QuoteNode(2:3))
) => UnitRange{Int64}
```

So I would like to understand better what is needed from the compiler.

**Reply by timholy** [reply]
@Tokazama, one way to better discover/document the need would be to change the definition to
```julia
julia> struct StaticInt{N} <: Integer
           n::Int
           StaticInt{N}() where {N} = new(N)
       end

julia> Base.convert(::Type{T}, i::StaticInt{N}) where {T<:Number,N} = convert(T, i.n)

julia> (::Type{T})(x::StaticInt{N}) where {T<:Integer,N} = T(x.n)
```
and see what fails in the ArrayInterface tests. I don't *think* the compiler will know that `n == N`, but of course I could be mistaken.

**Reply by AriMKatz** [reply]
@chriselrod iirc you found constant prop to be finicky. Can you explain exactly what you meant? 

 In general there's a question about how to guarantee optimizations, especially when composing code. I think expressing intent and invariants in  type information is one way to do that. Is the compiler ever going to be sufficiently good that we don't have to worry about corner cases? If not, is there some feature that we can use the type system like this without abusing it (like long compile time, 5 min or more for loopvec).

Maybe that belongs in a broader discussion and this specific issue is more  amenable to compiler optimizations

**Reply by chriselrod** [reply]
> @chriselrod iirc you found constant prop to be finicky. Can you explain exactly what you meant?

ArrayInterface has a lot of tests using `@inferred`. It can be painful/time consuming to figure out a way to get them to pass.

**Reply by chriselrod** [reply]
I often have to resort to `@generated`, e.g. https://github.com/JuliaArrays/ArrayInterface.jl/pull/245/commits/d497de8829b6a493506847d5a38bbac8fd6c2e5a is the most recent example.
Although that example requires Julia 1.6. It already inferred on 1.7 and 1.8 before the fix, meaning it isn't exactly an example someone from the compiler team can look at to fix.

**Reply by chriselrod** [reply]
If you do want an example from Julia nightly, ArrayInterface tests are currently failing there:
```julia
 dimnames: Error During Test at /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:96
  Test threw exception
  Expression: #= /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:96 =# @inferred(ArrayInterface.has_dimnames(x)) == true
  return type True does not match inferred return type Union{False, True}
  Stacktrace:
   [1] error(s::String)
     @ Base ./error.jl:35
   [2] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:464 [inlined]
   [3] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:96 [inlined]
   [4] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [5] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:91 [inlined]
   [6] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [7] top-level scope
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:8
dimnames: Error During Test at /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:98
  Test threw exception
  Expression: #= /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:98 =# @inferred(ArrayInterface.has_dimnames(ones(2, 2))) == false
  return type False does not match inferred return type Union{False, True}
  Stacktrace:
   [1] error(s::String)
     @ Base ./error.jl:35
   [2] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:464 [inlined]
   [3] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:98 [inlined]
   [4] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [5] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:91 [inlined]
   [6] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [7] top-level scope
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:8
dimnames: Error During Test at /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:99
  Test threw exception
  Expression: #= /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:99 =# @inferred(ArrayInterface.has_dimnames(Array{Int, 2})) == false
  return type False does not match inferred return type Union{False, True}
  Stacktrace:
   [1] error(s::String)
     @ Base ./error.jl:35
   [2] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:464 [inlined]
   [3] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:99 [inlined]
   [4] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [5] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:91 [inlined]
   [6] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [7] top-level scope
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:8
dimnames: Error During Test at /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:100
  Test threw exception
  Expression: #= /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:100 =# @inferred(ArrayInterface.has_dimnames(typeof(x))) == true
  return type True does not match inferred return type Union{False, True}
  Stacktrace:
   [1] error(s::String)
     @ Base ./error.jl:35
   [2] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:464 [inlined]
   [3] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:100 [inlined]
   [4] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [5] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:91 [inlined]
   [6] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [7] top-level scope
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:8
dimnames: Error During Test at /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:101
  Test threw exception
  Expression: #= /home/runner/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:101 =# @inferred(ArrayInterface.has_dimnames(typeof(view(x, :, 1, :)))) == true
  return type True does not match inferred return type Union{False, True}
  Stacktrace:
   [1] error(s::String)
     @ Base ./error.jl:35
   [2] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:464 [inlined]
   [3] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:101 [inlined]
   [4] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [5] macro expansion
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:91 [inlined]
   [6] macro expansion
     @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.9/Test/src/Test.jl:1357 [inlined]
   [7] top-level scope
     @ ~/work/ArrayInterface.jl/ArrayInterface.jl/test/dimensions.jl:8
```
several of the inference tests are failing there (but pass on Julia 1.6 and 1.7).
Relying on inference/const prop often feels like asking for trouble/heavy maintenance burden due to needing to check and fix things on each new release.
Relying more on `StaticInt`s is worse for compile times (assuming const prop works instead of causing inference failures that'd then make not using `StaticInt` worse for compile and runtime), but generally seems more robust.

**Reply by AriMKatz** [reply]
> Relying more on StaticInts is worse for compile times (assuming const prop works instead of causing inference failures that'd then make not using StaticInt worse for compile and runtime), but generally seems more robust.


Yeah so this is a good example of what I was talking about. If the compiler is never going to be as reliable, and they're always going to be new weird type domain / static computations that soss and catlab etc are doing, is there some generic solution to doing this without blowing up compile times or without having to write your own walled garden DSL and compiler?

**Reply by timholy** [reply]
Some of these problems can be solved by moving the tests into Julia's own test suite...that way, people working to improve inference will get immediate feedback about having made your life worse, and make fixes before merging to `master`. Since 1.9 is starting, one could take the best results obtained across 1.6-1.8 and commit them, using `@test_broken` if necessary to flag them as a challenge for future inference improvements.

I'm not saying this to try to discourage migration of StaticInt into Base, but a clear vision about what does and doesn't work can serve as a roadmap for future work that may have widespread benefits.

**Reply by Tokazama** [reply]
> > will the compiler turn [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1143) into [this line](https://github.com/JuliaLang/julia/blob/02abca3c293abdded1b3fa3629ced944c7a4e27b/base/range.jl#L1140)?
> 
> I don't think that specifically could happen, since the compiler cannot switch a `UnitRange` to a `OneTo`. But, it can constant propagate it:
> 
> ```
> julia> f() = intersect(1:3, 2:4)
> 
> julia> @code_typed f()
> CodeInfo(
> 1 ─     return $(QuoteNode(2:3))
> ) => UnitRange{Int64}
> ```
> 
> So I would like to understand better what is needed from the compiler.

If we are close to a point where we can just remove `OneTo` entirely because the compiler can elide related computations, then we don't really need `StaticInt`. However, my understanding is we will still need to store size in the type domain for things like `SArray` and find ways of propagating that information between array methods.

So it's less about the compiler being able to convert between types than it is about whether we can get rid of these highly specific types entirely because the compiler will infer related optimizations.
